rust/lance-io/src/scheduler.rs

```diff
@@ -0,0 +1,577 @@
+                idx: task_idx as u32,
+            }));
+        })
+    }
+
+    async fn do_indirect(
+        reader: Arc<dyn Reader>,
+        offsets_range: Range<u64>,
+        data_offset: u64,
+        data_idx: u32,
+        offset_idx: u32,
+    ) -> Result<DeliveredIndirect> {
+        let bytes = reader
+            .get_range(offsets_range.start as usize..offsets_range.end as usize)
+            .await?;
+        let length = bytes.len() / size_of::<u32>();
+        let values = ScalarBuffer::new(Buffer::from_bytes(bytes.into()), 0, length);
+        let offsets = UInt32Array::new(values, None);
+
+        let start = offsets.value(0);
+        let end = offsets.value(offsets.len() - 1);
+        let start = start as usize + data_offset as usize;
+        let end = end as usize + data_offset as usize;
+        let data_bytes = reader.get_range(start..end).await?;
+
+        Ok(DeliveredIndirect {
+            data_idx,
+            data: data_bytes,
+            offset_idx,
+            offsets,
+        })
+    }
+
+    fn spawn_indirect(&self, offset_idx: usize) -> JoinHandle<()> {
+        let data_idx = self.request.direct_requests.len() + offset_idx;
+        let task = &self.request.indirect_requests[offset_idx];
+
+        let reader = self.reader.clone();
+        let dest = self.dest.as_ref().unwrap().clone();
+        let offsets_range = task.offsets_range.clone();
+        let data_offset = task.data_offset;
+
+        tokio::spawn(async move {
+            let to_deliver = Self::do_indirect(
+                reader,
+                offsets_range,
+                data_offset,
+                data_idx as u32,
+                offset_idx as u32,
+            )
+            .await;
+            let mut dest = dest.lock().unwrap();
+            dest.deliver_indirect(to_deliver);
+        })
+    }
+
+    // We don't want to allocate a task as soon as it is created so this is
+    // deferred.  This should be called once a task is ready to start working
+    // and before it is consumed as a stream.
+    fn allocate(&mut self) {
+        let num_data_buffers =
+            self.request.direct_requests.len() + self.request.indirect_requests.len();
+        let data_buffers = vec![Bytes::default(); num_data_buffers];
+
+        let num_offset_buffers = self.request.indirect_requests.len();
+        let default_arr = UInt32Array::new_null(0);
+        let offset_buffers = vec![default_arr; num_offset_buffers];
+
+        self.dest = Some(Arc::new(Mutex::new(MutableBatch {
+            data_buffers,
+            offset_buffers,
+            when_done: self.when_done.take(),
+            err: None,
+        })));
+    }
+}
+
+impl Stream for IoTask {
+    type Item = JoinHandle<()>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        // This task_index is wrong because it doesn't take into account multiple buffers
```

Comment:
I this a problem?

Reply:
Oops :).  I ended up addressing this TODO and it is no longer relevant.  I'll remove it.